### ISSUE TITLE
Is this library prone to React optimization bug?

### ISSUE THREAD

**Issue by halt-hammerzeit** [erikras/redux-form#1705]
At first it seems that it's not and that it handles React reusing elements correctly:

``` js
    componentWillReceiveProps(nextProps) {
      if (this.props.name !== nextProps.name) {
        ...
        // unregister old name
        this.context._reduxForm.unregister(this.props.name)
        // register new name
        this.context._reduxForm.register(nextProps.name)
      }
    }

    componentWillUnmount() {
      this.context._reduxForm.unregister(this.name)
    }
```

But what if you had this form:
- `name`
- `email`
- `password`

And then it became this form:
- `email`
- `password`

What happens is:
- `name` field is transformed into `email` field: `name` is unregistered, `email` is registered
- `email` field is transformed into `password` field: `email` is unregistered, `password` is registered
- `password` field is unmounted: `password` is unregistered

And there you have it: no fields are registered though it's clearly nonsense.

**Comment by huan086** [comment]
Related: https://github.com/erikras/redux-form/issues/2016

**Comment by erikras** [comment]
Fix released as [`v6.5.0`](https://github.com/erikras/redux-form/releases/tag/v6.5.0).
